geopandas/plotting.py

```diff
@@ -516,31 +517,41 @@ def plot_dataframe(
     if values.dtype is np.dtype("O"):
         categorical = True
 
+    nan_idx = np.array(
+        [True if str(n) == "nan" or n is None else False for n in values]
+    )
+
     # Define `values` as a Series
     if categorical:
         if cmap is None:
             cmap = "tab10"
-        categories = list(set(values))
+        categories = list(set(values[~nan_idx]))
         categories.sort()
         valuemap = dict((k, v) for (v, k) in enumerate(categories))
-        values = np.array([valuemap[k] for k in values])
+        values = np.array([valuemap[k] for k in values[~nan_idx]])
 
     if scheme is not None:
         if classification_kwds is None:
             classification_kwds = {}
         if "k" not in classification_kwds:
             classification_kwds["k"] = k
 
-        binning = _mapclassify_choro(values, scheme, **classification_kwds)
+        binning = _mapclassify_choro(
+            values[~np.isnan(values)], scheme, **classification_kwds
+        )
         # set categorical to True for creating the legend
         categorical = True
-        binedges = [values.min()] + binning.bins.tolist()
+        binedges = [values[~np.isnan(values)].min()] + binning.bins.tolist()
         categories = [
             "{0:.2f} - {1:.2f}".format(binedges[i], binedges[i + 1])
             for i in range(len(binedges) - 1)
         ]
         values = np.array(binning.yb)
 
+    if categorical:
+        for n in np.where(nan_idx)[0]:
+            values = np.insert(values, n, values[0])
```

Comment:
Can you add some comments here? 
You are adding back values that correspond to the original values? (because those where removed when doing the conversion to integers?) And why the inserting the first value? (just as a placeholder?)

Could this be moved to the previous `if categorical` block? (I would think you cannot have `categorical` and `scheme` at the same time, but we don't check for that?)

Reply:
Above, in `if categorical` and `if scheme` block, I remove NaNs from values to set categories without them (for both categorical categories and for binning categories). Here I add placeholders to original positions to map values properly onto geometries later.

It could not be moved to categorical block above as it takes care of scheme one as well (scheme sets categorical to True).

Added comment to explain what is going on.